[all]: Added vector ADT and parsing in initial state.

lukeg101 opened the pull request:
This patch upstreams support for the sized-bounded vector type
in Herd, jingle, etc... unittests are included for completeness.

Specifically this patch adds parser support for vectors in the initial
state of litmus tests, and a corresponding type underneath so we can
write e.g:

```
int64_t v[2] = {1, 2}
```
which is expanded to v[0] = 1; v[1] = 2; etc...

Checks are added in the parser and elaboration stage in order
to ensure the size matches the supplied arguments.

Open questions (to fix before we merge):
- There is a conflict involving the syntax and semantics of
  the vector type. See test A130.litmus for a full explanation

in short: there is a type in herd that uses C list syntax we want:
```
v[8] = 3;    (* this is v+8 bytes*)
```
however the semantics we need is used by another type:
```
v+1 = 3; (* offset by the size of the type, so when uint64_t v, v+1 is
the same as v+8 bytes*)
```
ie we want the syntax of the first one but the semantics of the second.
These are 2 different types in Herd.

One Solution: We accept the first form so we have the C-list syntax, but
all offsets are byte offsets, and not the size of the type. E.g.
```
uint64 v[4] = { 1,2,3,4 }
P0;
(* program ...*)
forall (v[8] = 2) (*rather than v[1] = 2 *)
```
- There is an issue with `is_non_mixed_symbol` which checks the
  alignment of symbolic values, this fails in the presence of vectors
which are elaborated to v+0, v+8 etc... - it checks only that the
alignment is with respect to one element. e.g v+8 fails to align to the
second element of uint64_t v[2] = { 1,2 } since it checks the size of
uint64_t v.

I have made this obvious during the build process with a Warning, but it
would be worth getting your thoughts as this affects -variant unaligned.
etc...

- There is a shift reduce conflict in the parser owing to the
  introduction of {} into the initial state. I had a look at
  this but need to fix this - any tips?

maranget commented:
Hi @lukeg101, I am afraid your PR suffers from the recent merge of PR #84 (commit ef79f222ed8ad77c79457c3c9c77a945f12bf9fb).

lukeg101 commented:
I just realized I lost a few things in that rebase, please give me a moment to fix

jalglave commented:
hi @maranget did you have any thoughts on the two points raised by @lukeg101? More specifically (quoting Luke):

1. syntax and semantics of vectors:
> there is a type in herd that uses C list syntax we want:
> 
> v[8] = 3;    (* this is v+8 bytes*)
> however the semantics we need is used by another type:
> 
> v+1 = 3; (* offset by the size of the type, so when uint64_t v, v+1 is
> the same as v+8 bytes*)
> ie we want the syntax of the first one but the semantics of the second.
> These are 2 different types in Herd.
 
2. parsing conflict:
> shift reduce conflict in the parser owing to the introduction of {} into the initial state

Thanks,
Jade

jalglave commented:
hi @lukeg101 please can you make sure to adapt your code to not rely on stdint, and use @eth-arm's homegrown unit library instead?
Thanks,
Jade

lukeg101 commented:
> hi @lukeg101 please can you make sure to adapt your code to not rely on stdint, and use @eth-arm's homegrown unit library instead?
> Thanks,
> Jade

Yep will do, this patch shouldn't need stdint, but I will make sure subsequent patches for e.g ldp/stp use it

jalglave commented:
hi @maranget please can we have your thoughts on the questions asked by Luke? I would like to integrate this PR relatively soon.
Thanks,
Jade

maranget commented:
@lukeg101 Hi, the shift/reduce conflicts are relatively easy to solve. It suffices to get rid of some obsolete notation. Patch attached.
[0001-all-Get-rid-of-obsolete-init-notation-x-.-for-x.patch.txt](https://github.com/herd/herdtools7/files/5575008/0001-all-Get-rid-of-obsolete-init-notation-x-.-for-x.patch.txt)

Notice that you can get a detailed explanation of conflicts  with menhir option `--explain`. In dune context the easiest way to proceed is to add the option inside the `lib/dune` file (there are already other options there).

maranget commented:
@lukeg101 As to the second question, may I suggest a radical solution : do not allow notaton `<name>[<int>] = ..` in init. Does  not the usual array notation `<type> <name>[<int>]  = { ...}` suffice ?

maranget commented:
Then, low level accesses should be scaled according to base type of arrays and be aligned.

maranget commented on herd/unittests/AArch64/A128.litmus:
```diff
@@ -0,0 +1,9 @@
+AArch64 A128
+(* Tests accessing a location - symbolic location *)
+{  0:X2=v; v[1] = 44; v[0] = 22;}
```
This init section puzzles me. I rather see init section as _global_ declarations (well up to register initialisations..) Hence writing `v[1] = 44` without allocating the space for array `v` looks disturbing to me.

From the  perspective of the litmus tool, the idea of reserving space in the init section may be more apparent. Writing `int x` tells litmus to reserve space for some int, writing `uint64_t x` to reserve 8-byte etc.

maranget commented on herd/unittests/AArch64/A128.litmus.expected:
```diff
@@ -0,0 +1,10 @@
+Ok
+Witnesses
+Positive: 1 Negative: 0
+Condition forall (0:X0=22 /\ v=22 /\ v+1=44)
```
Something may have to change here, in the printing of the condition. The variable `v` being an array =, one expect to see `v[0]` and `v[1]` and not `v` and `v+1`.  Of course for `x` declared as a scalar `int x=0`, we do not want to have it printed as `v[0]`. Hence printing of conditions and outcome would depend upon types. This is manageable, I hope. I see no big difficulty for herd (just find the type in some environment) or litmus (which already relies much on types).  More so for tools such as mcompare...

maranget replied:
BTW: We already have a problem which could be solve by considering types while printing values. Ccnsider:
```
AArch64 T
{
int x=-1;
0:X1=x;
}
 P0          ;
LDR W0,[X1]  ;
locations [0:X0;]
````
Result is `0:X0=4294967295;`, which is correct in some sense. In fact the integer is printed as it is internally (`Int64.t`), while one should ignore the high order half and print it as a signed int (the defaut type). While litmus does correctly.

maranget commented on herd/unittests/AArch64/A130.litmus:
```diff
@@ -0,0 +1,28 @@
+  LDR X0, [X2, #16];
+
+forall (0:X0 = 3)
+
```
Hum this test may also result in a failure, as the space for `v[16]` is not allocated...

lukeg101 commented:
Hi Luc, I will reply to the above here before I submit an updated patch as there are some things I need you opinion on before I submit an updated PR.

> @lukeg101 As to the second question, may I suggest a radical solution : do not allow notation(*) `<name>[<int>] = ..` in init. Does not the usual array notation `<type> <name>[<int>] = { ...}` suffice ?
> 

This works for me, If I do not use the obsolete notation `<name>[<int>] = ..` in the initial state and apply your patch then the shift/reduce warning goes away. If we keep this patch, I can add you to `Co-Authored-By` in the commit message.

I will change the end-to-end tests to use `<type> <name>[<int>] = { ... }` instead

Do you want to prohibit the syntax `<name>[<int>] = ..` and throw an error? I can add that but it will make the parser more complex as we need to support other mechanisms such as MTE and Morello which use it. I do not know if this will have unintended side-effects - maybe we should leave this for now as this is a controversial patch already.

 > Something may have to change here, in the printing of the condition. The variable `v` being an array =, one expect to see `v[0]` and `v[1]` and not `v` and `v+1`. Of course for `x` declared as a scalar `int x=0`, we do not want to have it printed as `v[0]`. Hence printing of conditions and outcome would depend upon types. This is manageable, I hope. I see no big difficulty for herd (just find the type in some environment) or litmus (which already relies much on types). More so for tools such as mcompare...

This is the syntax versus semantics problem: the type `Location_global` prints as `v+0`, `v+1`... and has the semantics we want whereas `Location_deref` prints with the list-index syntax `v[0]` and `v[8]` but is byte offset (assuming `sizeof(v) == 8 bytes`).

I agree that the correct notation is `v[0]` and `v[1]` but this vector patch expands `int v = {1,2}` into multiple `Location_global` types  as `Location_global` has the semantics we want. This means the final state currently prints `v+0`, `v+1` etc... This allows vectors to work automatically with e.g mixed-mode without any new data types or strange translations to/from vectors to `Location_global`inside herd - we should not need strange translations here.

> BTW: We already have a problem which could be solved by considering types while printing values. Ccnsider:

Indeed this is present in mixed-mode tests too, see (A14.litmus linked to open PR [here](https://github.com/herd/herdtools7/commit/e9533a3dd3ca98e2fc679b649aa2e2df2f5ffc5a))

> * If this notation was allowed before your PR, it is not used in init anyway.

Do we use rely on`v[n]` or `Location_deref` syntax in the final state? this patch removes the `loc_deref` rule from `lib/stateParser` and replaces it with `Location_global` but if we do not ever use `loc_deref` then I can change the pretty printing of `Location_global` so that it has the syntax `v[0]` `v[1]` we want. This means things like `-debug mem` would also print `v[1] = ...` rather than `v+8` as it currently does - is this ok?

This solves our problems but prevents us from parsing `Location_deref` and changes the semantics of any previous tests using `v[...]` in the final state from `Location_deref` to `Location_global`.

This means we change the semantics of tests we accepted before, if we relied on `v[n]` in the final state. ie if there are previous tests that need `v[0]` and `v[8]` in the final state then these now have different meaning... this could break something if we miss it.  This is a semantics-breaking change, which is only safe is we are certain we do not need `Location_deref`

I worry about backwards compatibility. It may also help solve "ignore the high order half and print it as a signed int (the default type)" in the future

What do you think? should I change the printing of `v+0`,`v+8` to e.g `v[0]`,`v[1]` for `Location_global`, making `Location_deref` non-parsable? do we rely on `v[n]` in the final state currently? if we do not use `Location_deref` in the final state so I can go ahead and make the pretty printer change

maranget commented:
> What do you think? should I change the printing of `v+0`,`v+8` to e.g `v[0]`,`v[1]` for `Location_global`, making `Location_deref` non-parsable? do we rely on `v[n]` in the final state currently? if we do not use `Location_deref` in the final state then I can go ahead and make the pretty printer change

My opinion: get rid of v[<int>] = .. (array cell initialisation) in init. I do not see any MTE test that would use this notation, for Morello I am not sure, please check.

I am not sure you should expand array initialisations to cell initialisations very soon. This can be done later while creating the initial write events (function `initwrites` in `eventMonads.ml`). It is mandatory to proceed that way if `<name>[<int>]` cannot appear in init, which is worth a try, IMHO.

In final conditions, the notation is allowed. Using `t[..]` in final condition only applies when t has been declared to be an array in the init section. Looking at `stateParser.mly` notation `<name>[<int>]` indeed translates to Locations_deref. It looks ok. This is "high-level" (C-like syntax).  Clearly, accessing the content of the cell requires scaling by base type size. The corresponding location in the abstract environment being something like Symbol (<name>, sz*<int>).

lukeg101 commented:
Hi Luc,
Sorry about the delay, this work took quite a lot longer than expected. 
I have updated the patchset with a second patch to address your comments :) 

commit message summary here:

This patch addresses feedback from #94

This patch is not quite ready yet, as mixed-mode for vectors does
not work, and we have a new form of test for 'deliberate errors'
(see A134 test)

As this commit is quite large so I have split it into first patch and
second.

In detail this patch:
- Rebase and rename tests in testsuite - all pass
- run `make versions` to check backwards compatibility - passes with no
  warnings
- applied Luc's obsolete syntax patch to fix the shift-reduce error in
  the parser
- modifies tests so they  use `v[n] = {1,2,3}` `v[0] = 1; v[1] = 2 ...` syntax
- remove unallocated arrays in tests, make test A134 deliberately fails with
unallocated array error. The test infrastructure can not handle expected failures yet.
I have spoken to eth-arm about this and will follow up this patch with a fix.
- Updated A08 test with new `v[n]` syntax.
- Add optional vector metadata to Symbolic type including size of primitive type,
and total-size of vector. The initial state constructs the metadata, and we use
it in the final state (see lib/constant). Alignment checks are added for
vector elements.

lukeg101 commented:
> I am not sure you should expand array initialisations to cell initialisations very soon. This can be done later while creating the initial write events (function `initwrites` in `eventMonads.ml`). It is mandatory to proceed that way if `<name>[<int>]` cannot appear in init, which is worth a try, IMHO.

I was able to do this in archExtra_herd which allows us to expand array initializations into cell initializations before any checks are made.

> In final conditions, the notation is allowed. Using `t[..]` in final condition only applies when t has been declared to be an array in the init section. Looking at `stateParser.mly` notation `<name>[<int>]` indeed translates to Locations_deref. It looks ok. This is "high-level" (C-like syntax). Clearly, accessing the content of the cell requires scaling by base type size. The corresponding location in the abstract environment being something like `Symbol (<name>, sz*<int>)`.

I extended the Symbolic type with an optional argument containing the primitive type of the array in bytes (e.g uint64_t = 8), and the total size of the array. Then the above cell initialization passes this metadata from the initial state to the final state in test_herd

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -348,8 +348,66 @@ module Make(C:Config) (I:I) : S with module I = I
+              (* we expand v[3] = {a,b,c} into v+0 = a; v+1 = b; v+2 = c*)
+              (* where 1 is the sizeof the underlying primitive type *)
+              (* e.g uint64_t -> 8 bytes, so the above is v, v+8, v+16 *)
+              let vs = match I.V.get_vec v with
+                | Constant.ConcreteVector (sz,vs) when sz == List.length vs -> vs
+                | _ -> assert false in
```
done

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -348,8 +348,66 @@ module Make(C:Config) (I:I) : S with module I = I
-        List.fold_left (fun st (loc,(_,v)) -> State.add loc v st)
+        List.fold_left
+          (fun st (loc,(t,v)) ->
+            if MiscParser.is_array t then begin
```
done

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -348,8 +348,66 @@ module Make(C:Config) (I:I) : S with module I = I
+                vs
+            end else
+              (* if we have a value, store it *)
+              try
+                begin match State.find loc st with
+                  | _  -> Warn.fatal
+                           "Address %s non-unique in init state"
+                           (dump_location loc)
+                end
+              with Not_found   -> State.add loc v st)
```
done

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -348,8 +348,66 @@ module Make(C:Config) (I:I) : S with module I = I
+                (fun st (loc,(_,v))->
+                   try
+                     begin match State.find loc st with
+                     | _  -> Warn.fatal
```
fixed

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -348,8 +348,66 @@ module Make(C:Config) (I:I) : S with module I = I
+              else assert false in
+              let array_prim = MiscParser.get_array_primitive_ty t in
+              let prim_sz = MachSize.nbytes (size_of array_prim) in
+              let vec_data = Some (prim_sz, List.length vs) in
```
fixed

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -348,8 +348,66 @@ module Make(C:Config) (I:I) : S with module I = I
+              List.fold_left
+                (fun st (loc,(_,v))->
+                   try
+                     begin match State.find loc st with
```
that's a nice syntactic alternative, added, thanks!

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -348,8 +348,66 @@ module Make(C:Config) (I:I) : S with module I = I
+              let vs = match I.V.get_vec v with
+                | Constant.ConcreteVector (sz,vs) when sz == List.length vs -> vs
+                | _ -> assert false in
+              let locval = if is_global loc then
+                 match global loc with
+                 | Some x -> x
+                 | _ -> assert false (* unreachable *)
+              else assert false in
```
done

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -348,8 +348,66 @@ module Make(C:Config) (I:I) : S with module I = I
 
+      (* Simplified typing, size only, integer types only *)
+
+      let size_of = function
+        | "atomic_t"
+        | "int"|"long"
+        | "int32_t"
+        | "uint32_t" ->  MachSize.Word
+        | "char"|"int8_t" |"uint8_t" -> MachSize.Byte
+        | "short" | "int16_t" | "uint16_t" -> MachSize.Short
+        | "int64_t" | "uint64_t" -> MachSize.Quad
+        | "int128_t" | "uint128_t" -> MachSize.S128
+        | "intprt_t" | "uintprt_t" -> I.V.Cst.Scalar.machsize (* Maximal size = ptr size *)
+        | t ->
+            Warn.fatal "Cannot find the size of type %s" t
```
moved to `lib/MiscParser`

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -378,7 +436,7 @@ module Make(C:Config) (I:I) : S with module I = I
 
       let look_size_location env loc = match loc with
-      | Location_global (I.V.Val (Constant.Symbolic ((s,_,_),0))) -> look_size env s
+      | Location_global (I.V.Val (Constant.Symbolic ((s,_,_,_),0))) -> look_size env s
```
I have changed the `Symbolic` type to use a record now

lukeg101 commented on herd/archExtra_herd.ml:
```diff
@@ -348,8 +348,66 @@ module Make(C:Config) (I:I) : S with module I = I
+              let vs = List.mapi
+                (fun i v ->
+                  Location_global
+                    (I.V.Val (Constant.Symbolic((I.V.pp false locval, None, 0, vec_data),
```
Now that `Symbolic` type uses a record, this looks simpler now

lukeg101 commented:
Hi Luc, here is an update on the vector work so far:
- Applied your patches from `no-deref` branch, got existing vector tests working
- prohibit `v[i]=scalar` syntax in init, only allow `v[i]={scalar_list}`- added expected-failure test 
- Change printing of `Symbolic` type to `v+i` when used internally, and `v[i]` (scaled) in init/final state
- changed `Symbolic` type to use a record data type - lots of changes but code is better now
- Addressed your code review comments in the PR
- rename test `A134.litmus.expected` to `A134.litmus.expected-failure` in line with PR #133

lukeg101 commented:
Hi Luc,
here is another update:
- rebased code onto master, all tests pass, `make versions` passes
- fix the relative paths of `A134` and `A175` - tests pass
- Added the tests `L006`, `L007`, `L008` you kindly provided in December last year - fixed the alignment check in `lib/constant` so that all of those tests now pass

lukeg101 commented:
Hi Luc,
I rebased this PR onto master fixing the conflicts so that all tests pass,I had to change the following:
- rename tests `L006`->`L010` and `L007` -> `L010`
- fix merge conflicts, in particular in `misc_to_size` in `herd/archExtra_herd`, made `TyArray` the 'natural' machine size like `TyDefPointer` and `TyArray` (from `int`) in a previous version

tested with:
- make versions - no warnings
- internal regression suite - no warnings/errors other than the expected failures accounted for

maranget commented:
Thanks, I wish to perform a few modifications before merging, such as having sizes in array element initial writes. The easiest way to do so is for me to create a new PR that will extend this one.

Thanks again.

maranget commented:
Extended as #140.
